src/devices/sound/ap2010pcm.cpp

```diff
@@ -120,60 +121,94 @@ void ap2010pcm_device::reg_w(offs_t offset, uint32_t data, uint32_t mem_mask)
 
 	m_stream->update();
 
-	uint16_t sample = 0;
 	switch (offset) {
 		case 0x4/4:
 			if ((data & 0x78) == 0x78) {
-				m_sample_rate = 8000 * (1 + ((data & 2) >> 1));
+				m_sample_rate = 8000 * (1 + BIT(data, 1));
 				m_stream->set_sample_rate(m_sample_rate);
 
 				// When a new stream starts, stop playback of previous stream
-				m_fifo = {};
+				m_fifo_size = 0;
+				m_fifo_head = 0;
+				m_fifo_tail = 0;
 
 				LOG("pcm stream start, rate = %d\n", m_sample_rate);
 			}
 			break;
 		case 0xc/4:
 			if (ACCESSING_BITS_16_31) {
-				sample = (data & 0xffff0000U) >> 16;
-				if (sample != 0) {
-					m_fifo.push(sample);
-				}
+				fifo_push((data & 0xffff0000U) >> 16);
 			}
 			if (ACCESSING_BITS_0_15) {
-				sample = (data & 0x0000ffffU);
-				if (sample != 0) {
-					m_fifo.push(sample);
-				}
+				fifo_push(data & 0x0000ffffU);
 			}
 			break;
 		// These samples are always played first
 		case 0x10/4:
 			if (ACCESSING_BITS_16_31) {
-				sample = (data & 0xffff0000U) >> 16;
-				if (sample != 0) {
-					m_fifo_fast.push(sample);
-				}
+				fifo_fast_push((data & 0xffff0000U) >> 16);
 			}
 			if (ACCESSING_BITS_0_15) {
-				sample = (data & 0x0000ffffU);
-				if (sample != 0) {
-					m_fifo_fast.push(sample);
-				}
+				fifo_fast_push(data & 0x0000ffffU);
 			}
 			break;
-		// Panning. FIXME: Identify bits for each channel
+		// Panning. TODO: Identify bits for each channel
 		case 0x14/4:
-			LOG("pcm pan = %08x", data);
+			LOG("pcm pan = %08x\n", data);
 			break;
 		// Volume control. When video output is disabled, it's possible to adjust volume
 		// using the 2 touch areas on the bottom-left of the Storyware. Range 0..345
 		case 0x18/4:
-			m_volume = ((data & 0x1ff00000U) >> 20) / 345.0f;
-			if (m_volume > 1.0f) {
-				m_volume = 1.0f;
-			}
-			LOG("pcm vol = %08x -> %d", data, m_volume);
+			m_volume = std::max(((data & 0x1ff00000U) >> 20) / 345.0f, 1.0f);
+			LOG("pcm vol = %08x -> %d\n", data, m_volume);
```

Comment:
Should be `std::min`, not `std::max` – sorry if I said the wrong one before, I’m pretty sleep-deprived at the moment.

Reply:
No problem. Yeah I was getting some louder samples.